### PR TITLE
Enable database lock for singleton metricscollector without using consul based lock

### PR DESCRIPTION
**Pull request by paltanmoy**
- Implemented database lock to achieve singleton for metricscolletor without using consul (Need to enable this feature by setting enable_db_lock to true)

- Added unit tests

- Modified Integration tests

### CODE REVIEW

**Comment by paltanmoy** [comment]
@boyang9527 We can have a shared lock tables across the components. But that requires a separate liquibase changelog file (as part of a separate dblock module)

**Comment by boyang9527** [comment]
@paltanmoy  I guess there are some misunderstanding... 

The current one  is a single lock implementation.  The table only serves one lock.  I am hoping there will be database impl of distributed locks with an abstraction interface, which can support creating multiple locks,  and also different lock types will be even better... but we can start with non-blocking mutex

File: src/autoscaler/metricscollector/cmd/metricscollector/main.go [hunk 1]
```diff
@@ -208,3 +284,16 @@ func generateGUID(logger lager.Logger) string {
...
 	return uuid.String()
 }
+
+func getOwner(logger lager.Logger, conf *config.Config) string {
+	var owner string
+	if strings.TrimSpace(os.Getenv("VCAP_APPLICATION")) != "" {
+		appEnv, _ := cfenv.Current()
+		owner = appEnv.ID
+		logger.Info("ownership found in VCAP_APPLICATION", lager.Data{"owner": owner})
+	} else if conf.DBLock.Owner != "" {
```
Comment (paltanmoy) [inline src/autoscaler/metricscollector/cmd/metricscollector/main.go]: @kanekoh It's required parameter only when conf.EnableDBLock is set to true.

Comment (kanekoh) [inline src/autoscaler/metricscollector/cmd/metricscollector/main.go]: I wonder that getOwner function is called when conf.EnableDBLock is set to true. 
Am I missing something?

File: src/autoscaler/metricscollector/cmd/metricscollector/main.go [hunk 2]
```diff
@@ -159,6 +176,65 @@ func main() {
 	logger.Info("exited")
 }
 
+func initDBLockRunner(conf *config.Config, logger lager.Logger, lockDB db.LockDB) ifrit.Runner {
+	dbLockMaintainer := ifrit.RunFunc(func(signals <-chan os.Signal, ready chan<- struct{}) error {
+		ttl := conf.DBLock.LockTTL
+		lockTicker := time.NewTicker(ttl)
+		readyFlag := true
+		owner := getOwner(logger, conf)
```
Comment (boyang9527) [inline src/autoscaler/metricscollector/cmd/metricscollector/main.go]: If it is not deployed as a cf app,  we can use IP address and process ID, instead of using the one in config

If the component is deployed as one instance of a cf app, this will have issue.

File: src/autoscaler/db/sqldb/lock_sqldb_test.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+		Context("when no lock owner exist", func() {
+			BeforeEach(func() {
+				newlock := &models.Lock{Owner: "123456", Ttl: testTTL}
+				err = ldb.Acquire(newlock)
+			})
+			It("should successfully acquire the lock", func() {
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
```
Comment (boyang9527) [inline src/autoscaler/db/sqldb/lock_sqldb_test.go]: what about test case when the lock owner exists?

File: src/autoscaler/db/sqldb/lock_sqldb_test.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+			Expect(err).NotTo(HaveOccurred())
+			cleanLockTable()
+		})
+
+		AfterEach(func() {
+			err = ldb.Close()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Context("when lock owned by an instance", func() {
```
Comment (boyang9527) [inline src/autoscaler/db/sqldb/lock_sqldb_test.go]: what about  nobody owns the lock, and the lock is owned by others.

File: src/autoscaler/db/sqldb/lock_sqldb_test.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+			Expect(err).NotTo(HaveOccurred())
+			cleanLockTable()
+		})
+
+		AfterEach(func() {
+			err = ldb.Close()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		Context("when owner instance wants to release the lock", func() {
```
Comment (boyang9527) [inline src/autoscaler/db/sqldb/lock_sqldb_test.go]: missing test cases when lock table is empty,  and the lock is owned by others

File: src/autoscaler/metricscollector/cmd/metricscollector/main.go [hunk 6]
```diff
@@ -159,6 +176,65 @@ func main() {
...
+				if isLockAcquired && readyFlag {
+					close(ready)
+					readyFlag = false
+					logger.Info("successfully-acquired-lock", lager.Data{"owner": owner})
+				}
+			}
+		}
+	})
+	return dbLockMaintainer
+}
```
Comment (boyang9527) [inline src/autoscaler/metricscollector/cmd/metricscollector/main.go]: if only  Lock() and Release() are used,  why we have 	following as Public
 Fetch() (*models.Lock, error)
 Acquire(lock *models.Lock) error
 Renew(owner string) error

Comment (boyang9527) [reply src/autoscaler/metricscollector/cmd/metricscollector/main.go]: by the way,  I guess we can define a lock interface, and then use DBLock as one of the implementation.

**Comment by paltanmoy** [comment]
@boyang9527 , @kanekoh  most of your review comments has been addressed in the updated PR. Please have a look.

File: src/autoscaler/sync/dblock.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+func NewDatabaseLock(logger lager.Logger) *DatabaseLock {
+	return &DatabaseLock{
+		logger: logger,
+	}
+}
+
+func (dblock *DatabaseLock) InitDBLockRunner(ttl time.Duration, owner string, lockDB db.LockDB) ifrit.Runner {
+	dbLockMaintainer := ifrit.RunFunc(func(signals <-chan os.Signal, ready chan<- struct{}) error {
+		lockTicker := time.NewTicker(ttl)
+		readyFlag := true
```
Comment (boyang9527) [inline src/autoscaler/sync/dblock.go]: a little bit misleading, perhaps initialize with false, when it is ready then assign true value

File: src/autoscaler/sync/dblock.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+			case <-signals:
+				dblock.logger.Info("received-interrupt-signal", lager.Data{"owner": owner})
+				lockTicker.Stop()
+				releaseErr := lockDB.Release(owner)
+				if releaseErr != nil {
+					dblock.logger.Error("failed-to-release-lock ", releaseErr)
+				} else {
+					dblock.logger.Info("successfully-released-lock", lager.Data{"owner": owner})
+				}
+				readyFlag = true
```
Comment (boyang9527) [inline src/autoscaler/sync/dblock.go]: why set readyFalg, given you immediately return nil

File: src/autoscaler/sync/dblock.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+			dblock.logger.Info("lock-acquired-in-first-attempt", lager.Data{"owner": owner, "isLockAcquired": isLockAcquired})
+			close(ready)
+			readyFlag = false
+		}
+		for {
+			select {
+			case <-signals:
+				dblock.logger.Info("received-interrupt-signal", lager.Data{"owner": owner})
+				lockTicker.Stop()
+				releaseErr := lockDB.Release(owner)
```
Comment (boyang9527) [inline src/autoscaler/sync/dblock.go]: don't need to release if the lock is not acquired.

File: src/autoscaler/sync/dblock.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+func NewDatabaseLock(logger lager.Logger) *DatabaseLock {
+	return &DatabaseLock{
+		logger: logger,
+	}
+}
+
+func (dblock *DatabaseLock) InitDBLockRunner(ttl time.Duration, owner string, lockDB db.LockDB) ifrit.Runner {
+	dbLockMaintainer := ifrit.RunFunc(func(signals <-chan os.Signal, ready chan<- struct{}) error {
+		lockTicker := time.NewTicker(ttl)
```
Comment (boyang9527) [inline src/autoscaler/sync/dblock.go]: should not we use LockRetryInterval instead of ttl ?

File: src/autoscaler/sync/dblock.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+				readyFlag = true
+				return nil
+
+			case <-lockTicker.C:
+				dblock.logger.Info("retry-acquiring-lock", lager.Data{"owner": owner})
+				lock := &models.Lock{Owner: owner, Ttl: ttl}
+				isLockAcquired, lockErr := lockDB.Lock(lock)
+				if lockErr != nil {
+					dblock.logger.Error("failed-to-acquire-lock", lockErr)
+					releaseErr := lockDB.Release(owner)
```
Comment (boyang9527) [inline src/autoscaler/sync/dblock.go]: why bothered by release the lock if lock failed?

File: src/autoscaler/db/sqldb/lock_sqldb.go [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+	})
+	return err
+}
+
+func (ldb *LockSQLDB) Lock(lock *models.Lock) (bool, error) {
+	ldb.logger.Debug("acquiring-lock", lager.Data{"Owner": lock.Owner})
+	isLockAcquired := true
+	err := ldb.transact(ldb.sqldb, func(tx *sql.Tx) error {
+		newLock := false
+		fetchedLock, err := ldb.fetch(tx)
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply src/autoscaler/db/sqldb/lock_sqldb.go]: done

File: src/autoscaler/db/sqldb/lock_sqldb.go [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+			}
+			return err
+		}()
+
+		// golang sql package does not always retry query on ErrBadConn
+		if attempts >= 2 || (err != driver.ErrBadConn) {
+			break
+		} else {
+			ldb.logger.Error("deadlock-transaction", err, lager.Data{"attempts": attempts})
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply src/autoscaler/db/sqldb/lock_sqldb.go]: for debugging purpose

File: src/autoscaler/metricscollector/config/config.go [hunk 14]
```diff
@@ -70,27 +70,35 @@ type LockConfig struct {
...
 }
 
 type Config struct {
-	Cf        cf.CfConfig     `yaml:"cf"`
-	Logging   LoggingConfig   `yaml:"logging"`
-	Server    ServerConfig    `yaml:"server"`
-	Db        DbConfig        `yaml:"db"`
-	Collector CollectorConfig `yaml:"collector"`
-	Lock      LockConfig      `yaml:"lock"`
+	Cf           cf.CfConfig     `yaml:"cf"`
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply src/autoscaler/metricscollector/config/config.go]: How about refactoring the config through a separate PR ?

File: src/autoscaler/sync/dblock.go (hunk 10 above)
Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: we are renewing the lock every ttl timeframe. Database lock does not have any retry interval attribute.

File: src/autoscaler/sync/dblock.go (hunk 7 above)
Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: the readyFlag is being used to indicate whether it's ready to get the lock, initially set to true to indicate that the instance is ready to get the lock, as soon as it got the lock we made it to false.

File: src/autoscaler/sync/dblock.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	return &DatabaseLock{
+		logger: logger,
+	}
+}
+
+func (dblock *DatabaseLock) InitDBLockRunner(ttl time.Duration, owner string, lockDB db.LockDB) ifrit.Runner {
+	dbLockMaintainer := ifrit.RunFunc(func(signals <-chan os.Signal, ready chan<- struct{}) error {
+		lockTicker := time.NewTicker(ttl)
+		readyFlag := true
+		if owner == "" {
```
[the remark this answers is not in the document]
Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: the generateGUID() has been modified to send empty string during uuid generation error.

File: src/autoscaler/sync/dblock.go (hunk 9 above)
Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: Release is a delete operation on owner id, will succeed only if lock exists.

File: src/autoscaler/sync/dblock.go (hunk 11 above)
Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: it might fail to renew the lock for some reason (VM Crash/shutdown etc). In that case rather than waiting for the instance, we should forcefully release the existing lock, so that other instance could get the lock.

File: src/autoscaler/db/sqldb/lock_sqldb.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (boyang9527) [reply src/autoscaler/db/sqldb/lock_sqldb.go]: sorry i mean why this indicates 'deadlock'

File: src/autoscaler/sync/dblock.go (hunk 10 above)
Comment (boyang9527) [reply src/autoscaler/sync/dblock.go]: will this cause loss of the lock? I am assuming we need to refresh the lock less than ttl 
why not we add this? like consul lock?

File: src/autoscaler/sync/dblock.go (hunk 7 above)
Comment (boyang9527) [reply src/autoscaler/sync/dblock.go]: it is confusing since there is also ready channel while it has different meaning... If so can we change the variable name?

File: src/autoscaler/sync/dblock.go (hunk 11 above)
Comment (boyang9527) [reply src/autoscaler/sync/dblock.go]: other instance can still get the lock after ttl ...so that is fine not to do it...

Comment (paltanmoy) [reply src/autoscaler/sync/dblock.go]: If we release the lock as and when it fails to renew, then it does not require to wait till the ttl time to release forcefully. Lock switch will be limited within retry-interval timeframe in this scenario.

**Comment by paltanmoy** [comment]
@kanekoh & @boyang9527 any more comments ?

**Comment by ghaih** [comment]
No more comments from our side.Thanks.
